<a href="https://colab.research.google.com/github/udaymehta5/flyrank/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/udaymehta5/flyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My Lane

**Lane:** Structured Content Archetype Clustering (Core lane)
**Data:** Warehouse / starter dataset

Grouping pages by structural traits (word count, position tier, content type)
into archetypes — not by topic/meaning. This is structural clustering,
not semantic clustering.

## 2. The Question

**Question:** Do pages cluster into distinct structural archetypes, and do
these archetypes differ in performance?

**Unit of analysis:** A page.

**Output:** Cluster labels + a performance/trait profile per cluster.

**Action:** Prioritize restructuring pages in low-performing archetypes.

**Cost of wrong call:** Wasted effort restructuring pages that didn't need it,
or missing ones that did.

**Why ML:** Archetypes emerge from multiple features interacting — no single
column or rule reveals them; needs unsupervised discovery.

## 3. Quick Look at the Data

In [39]:
!git clone https://github.com/udaymehta5/flyrank.git
%cd flyrank

Cloning into 'flyrank'...
remote: Enumerating objects: 129, done.
remote: Counting objects: 100% (129/129), done.
remote: Compressing objects: 100% (85/85), done.
remote: Total 129 (delta 42), reused 93 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (129/129), 1.85 MiB | 12.94 MiB/s, done.
Resolving deltas: 100% (42/42), done.
/content/flyrank/flyrank/flyrank


In [40]:
import os

for root, dirs, files in os.walk("data"):
    for f in files:
        print(os.path.join(root, f))

data/raw/content_refresh_anonymized.csv


In [41]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print(f"Shape: {df.shape}")
print(f"Columns: {df.columns.tolist()}")

Shape: (30000, 44)
Columns: ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [42]:
combo_count = df.groupby(["content_type", "position_tier"]).ngroups
total_pages = len(df)
wc_std = df["word_count"].std()
wc_mean = df["word_count"].mean()

print(f"Unique content_type x position_tier combos: {combo_count}")
print(f"Total pages: {total_pages}")
print(f"Word count — mean: {wc_mean:.1f}, std: {wc_std:.1f}")
print(df["content_type"].value_counts())

print("\n--- Copy the line below into Section 3 markdown ---\n")
print(f"{combo_count} combinations across {total_pages} pages, with std of {wc_std:.1f} in word count — enough structural variety to support clustering.")

Unique content_type x position_tier combos: 14
Total pages: 30000
Word count — mean: 3107.8, std: 1452.4
content_type
keyword article       27207
feedly article         2096
comparison article      697
Name: count, dtype: int64

--- Copy the line below into Section 3 markdown ---

14 combinations across 30000 pages, with std of 1452.4 in word count — enough structural variety to support clustering.


 14 combinations across 30000 pages, with std of 1452.4 in word count — enough structural variety to support clustering.

## 4. Careful Words

- Clusters show correlation, not causation.
- Not semantic — says nothing about content meaning/topic.
- Cluster boundaries may shift with more features or full warehouse data.

In [43]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 5. Self-Check

- [done] Chose a predefined lane.
- [done] Named the decision and action.
- [done] Showed 2+ real numbers from starter data.
- [done] Explained why it's not just "train a model."
- [done] Avoided overclaiming causation or semantic meaning.